File: 007/solution_007.erl
-module(solution_007).
-export([start/0]).


is_divisible(Number)
        when Number rem 2 =:= 0 ->
            true;
is_divisible(Number) ->
    MaxDivisor = round(math:sqrt(Number)),
    Divisors = lists:seq(3, MaxDivisor, 2),
    is_divisible(Number, Divisors).

is_divisible(_, []) -> false;
is_divisible(Number, [Divisor|RemainingDivisors]) ->
    if
        Number rem Divisor == 0 ->
            true;
        true ->
            is_divisible(Number, RemainingDivisors)
    end.


nth_prime(N) ->
    nth_prime(N, 2, 3, 5).

nth_prime(N, PrimesFound, LastPrime, _)
        when PrimesFound >= N ->
            LastPrime;
nth_prime(N, PrimesFound, LastPrime, Count) ->
    case is_divisible(Count) of
        true -> nth_prime(N, PrimesFound, LastPrime, Count + 2);
        false -> nth_prime(N, PrimesFound + 1, Count, Count + 2)
    end.


solution(N) ->
    nth_prime(N).


start() ->
    io:format("~p\n", [solution(10001)]).

File: 007/solution_007.ml
let rec is_divisible = function
    | n, d when d >= n -> false
    | n, d when (n mod d == 0) -> true
    | n, d -> is_divisible (n, d+2)

let is_prime = function
    | 0 -> false
    | 1 -> false
    | 2 -> true
    | n when (n mod 2 == 0) -> false
    | n when is_divisible (n, 3) -> false
    | _ -> true

let nth_prime n =
    let rec nth_prime = function
    | 0, p, c -> p
    | n, p, c when is_prime (c) -> nth_prime (n-1, c, c+2)
    | n, p, c -> nth_prime (n, p, c+2)
    in
    nth_prime (n-2, 3, 5)

let project_euler_007 n =
    let solution = nth_prime n in
    print_newline (print_int solution)

let () = project_euler_007 10_001
